switchboard/tests/test_manager.py

```diff
@@ -679,6 +679,53 @@ def test_child_condition_including_parent_wins(self):
 
         assert_false(self.operator.is_active('test:child'))
 
+    def test_version_switch_no_user(self):
+        switch = Mock()
+        switch.save_version = Mock()
+        self.operator.context = dict()
+        self.operator.version_switch(switch)
+        switch.save_version.assert_called_with(username='')
+
+    def test_version_switch_user_dict(self):
+        username = 'test'
+        switch = Mock()
+        switch.save_version = Mock()
+        self.operator.context = dict(user=dict(username=username))
+        self.operator.version_switch(switch)
+        switch.save_version.assert_called_with(username=username)
+
+    def test_version_switch_user_class(self):
+        username = 'test'
+        switch = Mock()
+        switch.save_version = Mock()
+        self.operator.context = dict(user=Mock(username=username))
+        self.operator.version_switch(switch)
+        switch.save_version.assert_called_with(username=username)
+
+    def test_version_switch_user_garbage(self):
+        switch = Mock()
+        switch.save_version = Mock()
+
+        # Test with a non-User class.
+        class NonUser:
+            pass
+        self.operator.context = dict(user=NonUser())
+        self.operator.version_switch(switch)
+        switch.save_version.assert_called_with(username='')
+
+        # Test with a non-user dict.
+        self.operator.context = dict(user=dict())
+        self.operator.version_switch(switch)
+        switch.save_version.assert_called_with(username='')
+
+    def test_version_switch_save_error(self):
+        switch = Mock()
+        switch.save_version = Mock()
+        switch.save_version.side_effect = Exception('Boom!')
+        self.operator.context = dict()
+        # Don't need to assert, just need to make sure things don't explode.
+        self.operator.version_switch(switch)
```

Comment:
I've written tests that assert log messages in the past. In that case, the log messages were being gobbled up and turned into a report about revenue, so it was absolutely critical that the correct messages were being produced. That said, testing them was a serious PITA. Test runners tend to muck around with/intercept logging, making it hard to assert. I agree it would be nice, but just not nice enough to warrant the pain :-)